### PR TITLE
Upgrade to Mongo Java Driver 3.2.1 and MongoJack 2.5.1

### PR DESCRIPTION
**Pull request by joschi**
This PR upgrades Graylog to Mongo Java Driver 3.2.1 and MongoJack 2.5.1.

This raises the system requirements for Graylog to MongoDB 2.4.x or newer but opens up the opportunity to use "advanced" features like aggregations, that have existed for quite some time in MongoDB.

Driver compatibility chart: https://docs.mongodb.org/ecosystem/drivers/driver-compatibility-reference/#java-driver-compatibility

* Ubuntu 14.04 LTS (trusty): http://packages.ubuntu.com/trusty/mongodb
* Debian Linux (jessie): https://packages.debian.org/jessie/mongodb
* Debian Linux (wheezy-backports): https://packages.debian.org/wheezy-backports/mongodb
* CentOS 6 and 7: No MongoDB packages (needed to use the official MongoDB repositories anyway)
* Any other major Linux distribution: https://docs.mongodb.org/manual/administration/install-on-linux/#recommended

### CODE REVIEW

**Comment by bernd** [comment]
Please add this to [UPGRADING.asciidoc](https://github.com/Graylog2/graylog2-server/blob/master/UPGRADING.asciidoc).

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java [hunk 1]
```diff
@@ -86,16 +86,18 @@ static DBCollection prepareCollection(final MongoConnection mongoConnection) {
 
         DBCollection coll = db.getCollection(COLLECTION_NAME);
 
+        /* FIXME: Not supported by Fongo
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: What's the plan with this? Can we just drop the warning? I don't know why it's a problem when the collection is capped.. :wink:

Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: This is actually an artifact of the first implementation of the Cluster Event feature which had been using a capped collection to be able to use a tailable cursor and automatically get rid of "old" events. The problem was, that a capped collection only allows fixed size documents, so updating the events document (e. g. the list of nodes that have already processed the event) wasn't possible.

This never made it into a release, so I think we can safely remove the check (I'd do this in this PR if we reach consensus about it). The alternatives would be to either disable the related tests or extend Fongo to support capped collections again (they were supported in the past).

See 8d4765a3c831260e18f8a5d4935b4025846cc573 for details.

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java [hunk 2]
```diff
@@ -86,16 +86,18 @@ static DBCollection prepareCollection(final MongoConnection mongoConnection) {
...
         }
+        */
 
         coll.createIndex(DBSort
                 .asc("timestamp")
                 .asc("producer")
                 .asc("consumers"));
 
-        coll.setWriteConcern(WriteConcern.FSYNCED);
+        coll.setWriteConcern(WriteConcern.JOURNALED);
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: We are using `FSYNCED` in some other places as well. Should we switch those as well? 
```
$ git grep -F WriteConcern.
graylog2-server/src/main/java/org/graylog2/cluster/ClusterConfigServiceImpl.java:        coll.setWriteConcern(WriteConcern.FSYNCED);
graylog2-server/src/main/java/org/graylog2/cluster/ClusterConfigServiceImpl.java:        dbCollection.update(DBQuery.is("type", canonicalClassName), clusterConfig, true, false, WriteConcern.FSYNCED);
graylog2-server/src/main/java/org/graylog2/database/MongoConnectionImpl.java:            db.setWriteConcern(WriteConcern.SAFE);
graylog2-server/src/main/java/org/graylog2/events/ClusterEventCleanupPeriodical.java:            final WriteResult<ClusterEvent, String> writeResult = dbCollection.remove(query, WriteConcern.FSYNCED);
graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java:        coll.setWriteConcern(WriteConcern.JOURNALED);
graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java:            final String id = dbCollection.save(clusterEvent, WriteConcern.JOURNALED).getSavedId();
graylog2-server/src/main/java/org/graylog2/metrics/MongoDbMetricsReporter.java:            collection.insert(docs, WriteConcern.UNACKNOWLEDGED);
graylog2-server/src/test/java/org/graylog2/cluster/ClusterConfigServiceImplTest.java:        assertThat(collection.getWriteConcern()).isEqualTo(WriteConcern.FSYNCED);
graylog2-server/src/test/java/org/graylog2/cluster/ClusterConfigServiceImplTest.java:        assertThat(collection.getWriteConcern()).isEqualTo(WriteConcern.FSYNCED);
graylog2-server/src/test/java/org/graylog2/events/ClusterEventPeriodicalTest.java:        assertThat(collection.getWriteConcern()).isEqualTo(WriteConcern.JOURNALED);
graylog2-server/src/test/java/org/graylog2/events/ClusterEventPeriodicalTest.java:        assertThat(collection.getWriteConcern()).isEqualTo(WriteConcern.JOURNALED);
```

**Comment by bernd** [comment]
I was testing this with MongoDB 2.0.4 and I didn't ran into any error or warning. I was able to create dashboards, streams, inputs. No problem. I the minimum requirement is actually 2.4, we warn the user or maybe even stop the server if it's connected to an older MongoDB version.

```
2016-02-15 10:31:05,827 INFO : org.mongodb.driver.cluster - Monitor thread successfully connected to server with description ServerDescription{address=127.0.0.1:27017, type=STANDALONE, state=CONNECTED, ok=true, version=ServerVersion{versionList=[2, 0, 4]}, minWireVersion=0, maxWireVersion=0, maxDocumentSize=16777216, roundTripTimeNanos=1896308}
```

**Comment by bernd** [comment]
Can you rebase against master?! Thanks! :smiley:

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java (hunk 2 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: Yes, I think replacing `FSYNCED` with `JOURNALED` is a good idea as it doesn't block a single server to fsync data to disk but makes sure that the document has been written to MongoDB's on disk journal.

* https://docs.mongodb.org/v3.2/reference/command/fsync/
* http://api.mongodb.org/java/3.2/com/mongodb/WriteConcern.html#FSYNCED
* http://api.mongodb.org/java/3.2/com/mongodb/WriteConcern.html#JOURNALED

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java (hunk 1 above)
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: I think we can remove the warning then.

File: graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java (hunk 2 above)
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/events/ClusterEventPeriodical.java]: :+1: Also, the documentation for `FSYNC` says: "prefer JOURNALED"

**Comment by bernd** [comment]
LGTM :+1:

**Pull request merged**